File: src/dere_discord/agent.py
"""Claude Agent integration for Discord messaging via centralized daemon."""

from __future__ import annotations

import asyncio
from collections import defaultdict
from collections.abc import Awaitable, Callable

from loguru import logger

from dere_shared.agent_models import SessionConfig, StreamEventType

from .daemon_agent import DaemonAgentClient
from .persona import PersonaProfile
from .session import SessionManager


class DiscordAgent:
    """Bridge Discord messages to Claude via the centralized daemon agent."""

    def __init__(
        self,
        sessions: SessionManager,
        daemon_client: DaemonAgentClient,
        *,
        context_enabled: bool = True,
    ):
        self._sessions = sessions
        self._daemon = daemon_client
        self._locks: defaultdict[str, asyncio.Lock] = defaultdict(asyncio.Lock)
        self._context_enabled = context_enabled

    async def _stream_response(
        self,
        session,
        send_text_message: Callable,
        send_tool_summary: Callable,
    ) -> tuple[bool, list[str], list[str], list[str]]:
        """Stream response from daemon and handle text/tool blocks."""
        pre_tool_chunks: list[str] = []
        post_tool_chunks: list[str] = []
        tool_events: list[str] = []
        tool_seen = False

        try:
            async for event in self._daemon.query(session.pending_prompt):
                if event.type == StreamEventType.TEXT:
                    text = event.data.get("text", "")
                    if text:
                        if tool_seen:
                            post_tool_chunks.append(text)
                        else:
                            pre_tool_chunks.append(text)

                elif event.type == StreamEventType.THINKING:
                    text = event.data.get("text", "")
                    if text:
                        if tool_seen:
                            post_tool_chunks.append(text)
                        else:
                            pre_tool_chunks.append(text)

                elif event.type == StreamEventType.TOOL_USE:
                    if not tool_seen:
                        tool_seen = True
                        initial_text = "".join(pre_tool_chunks).strip()
                        if initial_text:
                            await send_text_message(initial_text, session.persona_profile)
                            await self._sessions.capture_message(
                                session, content=initial_text, role="assistant"
                            )
                        pre_tool_chunks.clear()

                    name = event.data.get("name", "unknown")
                    tool_input = event.data.get("input", {})
                    preview = self._preview(tool_input)
                    tool_events.append(f"Running `{name}`: {preview}")

                elif event.type == StreamEventType.TOOL_RESULT:
                    output = event.data.get("output", "")
                    is_error = event.data.get("is_error", False)
                    formatted = self._preview(output, limit=400)
                    if is_error:
                        tool_events.append(f"Tool failed\n{formatted}".strip())
                    else:
                        tool_events.append(f"Tool completed\n{formatted}".strip())

                elif event.type == StreamEventType.ERROR:
                    msg = event.data.get("message", "Unknown error")
                    logger.error("Agent error: {}", msg)
                    if not event.data.get("recoverable", True):
                        break

                elif event.type == StreamEventType.DONE:
                    break

        except Exception:
            logger.exception("Failed while streaming response")
            raise

        return tool_seen, pre_tool_chunks, post_tool_chunks, tool_events

    async def _finalize_response(
        self,
        session,
        tool_seen: bool,
        pre_tool_chunks: list[str],
        post_tool_chunks: list[str],
        tool_events: list[str],
        send_text_message: Callable,
        send_tool_summary: Callable,
    ) -> None:
        """Finalize response by sending remaining text and tool summaries."""
        persona_profile = session.persona_profile

        if not tool_seen:
            response_text = "".join(pre_tool_chunks).strip()
            await send_text_message(response_text, persona_profile)
            if response_text:
                await self._sessions.capture_message(
                    session, content=response_text, role="assistant"
                )
        else:
            final_text = "".join(post_tool_chunks).strip()

            if tool_events:
                await send_tool_summary(tool_events, persona_profile)

            if final_text:
                await send_text_message(final_text, persona_profile)
                await self._sessions.capture_message(
                    session, content=final_text, role="assistant"
                )

        await self._sessions.schedule_summary(session)

    async def handle_message(
        self,
        *,
        guild_id: int | None,
        channel_id: int,
        user_id: int | None,
        content: str,
        send_initial: Callable[[], Awaitable[None]],
        send_text_message: Callable[[str, PersonaProfile], Awaitable[None]],
        send_tool_summary: Callable[[list[str], PersonaProfile], Awaitable[None]],
        finalize: Callable[[], Awaitable[None]],
    ) -> None:
        """Handle a Discord user message and stream response back.

        Callbacks:
            send_initial() -> Awaitable[None]                  - invoked before streaming begins.
            send_text_message(text, profile) -> Awaitable[None] - send assistant text message.
            send_tool_summary(events, profile) -> Awaitable[None] - send tool activity summary.
            finalize() -> Awaitable[None]                      - cleanup (e.g., stop typing).
        """
        key = self._make_key(guild_id, channel_id)
        lock = self._locks[key]

        async with lock:
            session = await self._sessions.ensure_session(
                guild_id=guild_id,
                channel_id=channel_id,
                user_id=user_id,
            )

            await self._sessions.capture_message(session, content=content, role="user")

            session.pending_prompt = content

            config = SessionConfig(
                working_dir=session.project_path,
                output_style="discord",
                personality=",".join(session.personas),
                user_id=session.user_id,
                include_context=self._context_enabled,
            )

            try:
                await self._daemon.ensure_session(
                    config=config,
                    session_id=session.daemon_session_id,
                )
                session.daemon_session_id = self._daemon.session_id
            except Exception:
                logger.exception("Failed to ensure daemon session")
                raise

            await send_initial()

            try:
                tool_seen, pre_tool_chunks, post_tool_chunks, tool_events = (
                    await self._stream_response(session, send_text_message, send_tool_summary)
                )
            except Exception:
                await finalize()
                raise
            else:
                await finalize()

                await self._finalize_response(
                    session,
                    tool_seen,
                    pre_tool_chunks,
                    post_tool_chunks,
                    tool_events,
                    send_text_message,
                    send_tool_summary,
                )

    def _make_key(self, guild_id: int | None, channel_id: int) -> str:
        return f"{guild_id or 'dm'}:{channel_id}"

    def _preview(self, data: object, *, limit: int = 120) -> str:
        if isinstance(data, str):
            text = data.strip()
            if not text:
                return "(no output)"
            if len(text) > limit:
                return text[: limit - 3] + "..."
            return text

        if isinstance(data, dict):
            for key in ("text", "content", "stdout", "output", "result", "command"):
                if key in data and data[key]:
                    return self._preview(data[key], limit=limit)
            return "(no output)"

        if isinstance(data, list | tuple):
            if not data:
                return "(no output)"
            parts = [self._preview(item, limit=limit) for item in data]
            text = "\n".join(part for part in parts if part and part != "(no output)")
            if not text:
                return "(no output)"
            if len(text) > limit:
                return text[: limit - 3] + "..."
            return text

        text = str(data).strip()
        if not text:
            return "(no output)"
        if len(text) > limit:
            return text[: limit - 3] + "..."
        return text

File: packages/discord/src/agent.ts
import type { AgentClient } from "@dere/daemon-client";
import type { PersonaProfile } from "./persona.js";
import type { ChannelSession, SessionManager } from "./session.js";

type Awaitable<T> = T | Promise<T>;

type SendTextMessage = (text: string, profile: PersonaProfile) => Promise<void>;
type SendToolSummary = (events: string[], profile: PersonaProfile) => Promise<void>;

type MessageCallbacks = {
  sendInitial: () => Awaitable<void>;
  sendTextMessage: SendTextMessage;
  sendToolSummary: SendToolSummary;
  finalize: () => Awaitable<void>;
};

export class DiscordAgent {
  private readonly sessions: SessionManager;
  private readonly daemon: AgentClient;
  private readonly locks = new Map<string, Promise<void>>();
  private readonly contextEnabled: boolean;

  constructor(sessions: SessionManager, daemon: AgentClient, contextEnabled = true) {
    this.sessions = sessions;
    this.daemon = daemon;
    this.contextEnabled = contextEnabled;
  }

  private async withLock<T>(key: string, fn: () => Promise<T>): Promise<T> {
    const previous = this.locks.get(key) ?? Promise.resolve();
    let release: () => void = () => undefined;
    const next = new Promise<void>((resolve) => {
      release = resolve;
    });
    this.locks.set(
      key,
      previous.then(() => next),
    );
    await previous;
    try {
      return await fn();
    } finally {
      release();
      if (this.locks.get(key) === next) {
        this.locks.delete(key);
      }
    }
  }

  private async streamResponse(
    session: ChannelSession,
    callbacks: { sendTextMessage: SendTextMessage; sendToolSummary: SendToolSummary },
  ): Promise<{ toolSeen: boolean; preTool: string[]; postTool: string[]; toolEvents: string[] }> {
    const preTool: string[] = [];
    const postTool: string[] = [];
    const toolEvents: string[] = [];
    let toolSeen = false;

    try {
      for await (const event of this.daemon.query(session.pendingPrompt)) {
        if (event.type === "text" || event.type === "thinking") {
          const text = typeof event.data.text === "string" ? event.data.text : "";
          if (text) {
            if (toolSeen) {
              postTool.push(text);
            } else {
              preTool.push(text);
            }
          }
        } else if (event.type === "tool_use") {
          if (!toolSeen) {
            toolSeen = true;
            const initialText = preTool.join("").trim();
            if (initialText) {
              await callbacks.sendTextMessage(initialText, session.personaProfile);
              await this.sessions.captureMessage(session, {
                content: initialText,
                role: "assistant",
              });
            }
            preTool.length = 0;
          }

          const name = typeof event.data.name === "string" ? event.data.name : "unknown";
          const toolInput = event.data.input ?? {};
          const preview = this.preview(toolInput);
          toolEvents.push(`Running \`${name}\`: ${preview}`);
        } else if (event.type === "tool_result") {
          const output = event.data.output ?? "";
          const isError = Boolean(event.data.is_error);
          const formatted = this.preview(output, 400);
          toolEvents.push(`${isError ? "Tool failed" : "Tool completed"}\n${formatted}`.trim());
        } else if (event.type === "error") {
          const message =
            typeof event.data.message === "string" ? event.data.message : "Unknown error";
          console.error(`Agent error: ${message}`);
          if (event.data.recoverable === false) {
            break;
          }
        } else if (event.type === "done") {
          break;
        }
      }
    } catch (error) {
      console.error(`Failed while streaming response: ${String(error)}`);
      throw error;
    }

    return { toolSeen, preTool, postTool, toolEvents };
  }

  private async finalizeResponse(
    session: ChannelSession,
    result: { toolSeen: boolean; preTool: string[]; postTool: string[]; toolEvents: string[] },
    callbacks: { sendTextMessage: SendTextMessage; sendToolSummary: SendToolSummary },
  ): Promise<void> {
    const { toolSeen, preTool, postTool, toolEvents } = result;
    const personaProfile = session.personaProfile;

    if (!toolSeen) {
      const responseText = preTool.join("").trim();
      await callbacks.sendTextMessage(responseText, personaProfile);
      if (responseText) {
        await this.sessions.captureMessage(session, {
          content: responseText,
          role: "assistant",
        });
      }
    } else {
      const finalText = postTool.join("").trim();
      if (toolEvents.length > 0) {
        await callbacks.sendToolSummary(toolEvents, personaProfile);
      }
      if (finalText) {
        await callbacks.sendTextMessage(finalText, personaProfile);
        await this.sessions.captureMessage(session, {
          content: finalText,
          role: "assistant",
        });
      }
    }

    await this.sessions.scheduleSummary(session);
  }

  async handleMessage(args: {
    guildId: string | null;
    channelId: string;
    userId: string | null;
    content: string;
    callbacks: MessageCallbacks;
  }): Promise<void> {
    const key = `${args.guildId ?? "dm"}:${args.channelId}`;
    await this.withLock(key, async () => {
      const session = await this.sessions.ensureSession({
        guildId: args.guildId,
        channelId: args.channelId,
        userId: args.userId ?? null,
      });

      await this.sessions.captureMessage(session, { content: args.content, role: "user" });
      session.pendingPrompt = args.content;

      const config = {
        working_dir: session.projectPath,
        output_style: "discord",
        personality: session.personas.join(","),
        user_id: session.userId ?? null,
        include_context: this.contextEnabled,
      };

      try {
        await this.daemon.ensureSession(config, session.daemonSessionId);
        session.daemonSessionId = this.daemon.sessionId;
      } catch (error) {
        console.error("Failed to ensure daemon session", error);
        throw error;
      }

      await args.callbacks.sendInitial();

      try {
        const result = await this.streamResponse(session, {
          sendTextMessage: args.callbacks.sendTextMessage,
          sendToolSummary: args.callbacks.sendToolSummary,
        });
        await args.callbacks.finalize();
        await this.finalizeResponse(session, result, {
          sendTextMessage: args.callbacks.sendTextMessage,
          sendToolSummary: args.callbacks.sendToolSummary,
        });
      } catch (error) {
        await args.callbacks.finalize();
        throw error;
      }
    });
  }

  private preview(data: unknown, limit = 120): string {
    if (typeof data === "string") {
      const text = data.trim();
      if (!text) {
        return "(no output)";
      }
      return text.length > limit ? `${text.slice(0, limit - 3)}...` : text;
    }

    if (data && typeof data === "object" && !Array.isArray(data)) {
      const record = data as Record<string, unknown>;
      for (const key of ["text", "content", "stdout", "output", "result", "command"]) {
        if (record[key]) {
          return this.preview(record[key], limit);
        }
      }
      return "(no output)";
    }

    if (Array.isArray(data)) {
      if (data.length === 0) {
        return "(no output)";
      }
      const parts = data.map((item) => this.preview(item, limit));
      const text = parts.filter((part) => part && part !== "(no output)").join("\n");
      if (!text) {
        return "(no output)";
      }
      return text.length > limit ? `${text.slice(0, limit - 3)}...` : text;
    }

    const text = String(data ?? "").trim();
    if (!text) {
      return "(no output)";
    }
    return text.length > limit ? `${text.slice(0, limit - 3)}...` : text;
  }
}
